packages/webapp/src/components/Task/AddSoilAmendmentProducts/ProductCard/ProductDetails.tsx

```diff
@@ -0,0 +1,266 @@
+  ) => void;
+  setFieldValidity: (isValid: boolean) => void;
+};
+
+const isNewProduct = (productId: ProductId): boolean => typeof productId === 'string';
+
+export const defaultValues = {
+  [FIELD_NAMES.SUPPLIER]: '',
+  [FIELD_NAMES.COMPOSITION]: {
+    [FIELD_NAMES.UNIT]: Unit.RATIO,
+    [FIELD_NAMES.N]: NaN,
+    [FIELD_NAMES.P]: NaN,
+    [FIELD_NAMES.K]: NaN,
+  },
+};
+
+const ProductDetails = ({
+  productId,
+  products = [],
+  isReadOnly,
+  isExpanded,
+  farm: { farm_id, country_id, interested },
+  expand,
+  unExpand,
+  toggleExpanded,
+  clearProduct,
+  setProductId,
+  onSave,
+  setFieldValidity,
+}: ProductDetailsProps) => {
+  const { t } = useTranslation();
+  const [isEditingProduct, setIsEditingProduct] = useState(false);
+  const previousProductIdRef = useRef<number | string | undefined>(productId);
+
+  const inCanada = country_id === CANADA;
+  const isDetailDisabled = isReadOnly || !isEditingProduct;
+  const isProductEntered = !!productId;
+
+  const {
+    control,
+    getValues,
+    handleSubmit,
+    reset,
+    setFocus,
+    trigger,
+    register,
+    formState: { errors, isValid, isDirty },
+  } = useForm<FormFields>({
+    mode: 'onBlur',
+    defaultValues,
+  });
+
+  useEffect(() => {
+    const selectedProduct = products.find(({ product_id }) => product_id === productId);
+    const { supplier, on_permitted_substances_list, n, p, k, npk_unit } = selectedProduct || {};
+
+    const wasAddingNewProduct = isNewProduct(previousProductIdRef.current);
+    const isAddingNewProduct = !!(productId && !selectedProduct);
+    const shouldNotResetFields = wasAddingNewProduct && isAddingNewProduct;
+
+    if (!productId || !shouldNotResetFields) {
+      reset({
+        [FIELD_NAMES.SUPPLIER]: supplier || '',
+        [FIELD_NAMES.PERMITTED]: on_permitted_substances_list || undefined,
+        [FIELD_NAMES.COMPOSITION]: {
+          [FIELD_NAMES.UNIT]: npk_unit || Unit.RATIO,
+          [FIELD_NAMES.N]: n ?? NaN,
+          [FIELD_NAMES.P]: p ?? NaN,
+          [FIELD_NAMES.K]: k ?? NaN,
+        },
+      });
+    }
+
+    setIsEditingProduct(isAddingNewProduct);
+    if (isAddingNewProduct) {
+      expand();
+    }
+
+    trigger();
+
+    if (isAddingNewProduct && productId) {
+      // Wait for the card to be expaneded
+      setTimeout(() => {
+        setFocus(FIELD_NAMES.SUPPLIER);
+      }, 0);
+    }
+    previousProductIdRef.current = productId;
+  }, [productId]);
+
+  useEffect(() => {
+    setFieldValidity(isValid);
+  }, [isValid]);
+
+  const onCancel = () => {
+    if (isNewProduct(productId)) {
+      clearProduct();
+      setProductId(undefined);
+      unExpand();
+    } else {
+      reset();
+      setIsEditingProduct(false);
+    }
+  };
+
+  const onSubmit = (data: FormFields) => {
+    const callback = isNewProduct(productId) ? setProductId : undefined;
+    onSave({ ...data, farm_id, product_id: productId, type: TASK_TYPES.SOIL_AMENDMENT }, callback);
```

Comment:
Clever! So this callback will get passed to both PATCH and POST mutations? It's a bit easier to imagine with Saga but I like how it's all set up.